datalad_next/iter_collections/annexworktree.py

```diff
@@ -0,0 +1,194 @@
+
+    Yields
+    ------
+    :class:`AnnexWorktreeItem` or `AnnexWorktreeFileSystemItem`
+    """
+    # get a lookup mapping platform-paths to annex key and bytesize
+    # for any path connected to an annex key
+    annex = _get_annexinfo(path)
+
+    # we use git-annex-examinekey for annex object location reporting.
+    # we cannot use git-annex-contentlocation, because it only reports on
+    # present annex objects, and here we also need to report on would-be
+    # locations
+    with annexjson_batchcommand(
+            ['git', 'annex', 'examinekey', '--json', '--batch'],
+            cwd=path,
+    ) as exkey:
+        for item in iter_gitworktree(
+            path=path,
+            untracked=untracked,
+            # we relay the link_target flag to ensure homogeneous behavior
+            # with iter_gitworktree(), but will also do further processing
+            # below
+            link_target=False,
+            # we need to account for annexed files and would rather open
+            # their content in the annex, so we have to do this separately
+            fp=False
+        ):
+            ainfo = annex.get(item.name, {})
+            akey = ainfo.get('key')
+            aobjpath = None
+            if link_target and akey:
+                keyprops = exkey(bytes(akey, encoding="utf-8") + b'\n')
+                aobjpath = keyprops.get('objectpath')
+                if aobjpath:
+                    aobjpath = PurePath(aobjpath)
+
+            asize = ainfo.get('bytesize')
+            if asize is not None:
+                asize = int(asize)
+            yield AnnexWorktreeItem(
+                name=item.name,
+                gitsha=item.gitsha,
+                gittype=item.gittype,
+                annexkey=ainfo.get('key'),
+                annexsize=asize,
+                annexobjpath=aobjpath,
+            )
+
+
+def _get_annexinfo(path: Path) -> dict:
+    from datalad.support.annexrepo import GeneratorAnnexJsonNoStderrProtocol
+
+    map = {
+        'bytesize': lambda x: x,
+        'key': lambda x: x,
+    }
+
+    with run(
+        [
+            'git', 'annex', 'find',
+            # we want everything regardless of local availability.
+            # the faster `--anything` came only in 10.20230126
+            #'--anything',
+            '--include=*',
+            '--json', '--json-error-messages',
+            '.',
+        ],
+        protocol_class=GeneratorAnnexJsonNoStderrProtocol,
+        # run in the directory we want info on
+        cwd=path,
+    ) as annexfind:
+        return {
+            # git-annex reports the file path in POSIX conventions,
+            # even on windows, but the hashdir report comes in
+            # platform conventions.
+            # we need to go to platform conventions to match the behavior of
+            # iter_gitworktree()
+            PurePath(PurePosixPath(r['file'])):
+            {
+                k: map[k](r[k]) for k in map.keys() if k in r
```

Comment:
`map.keys()` is not required, `for k in map` would iterate over all keys. Also, instead of mapping a key to a value, i.e. `map[k]`, one could use `items()`. e.g.:

```suggestion
                k: m(r[k]) for k, m in map.items() if k in r
```